### PR TITLE
Add command to retrieve IDevID CSR from persistent storage.

### PR DESCRIPTION
**Pull request by clundin25**
* This resolves https://github.com/chipsalliance/caliptra-sw/issues/1687, for ROM.

### ROM Backwards compatibility

If the runtime is used with a ROM that does not support CSR generation, `RUNTIME_GET_IDEV_ID_UNSUPPORTED_ROM` is returned as an error.

This was tested manually by pulling in the ROM checked into https://github.com/chipsalliance/caliptra-sw/releases/tag/release_v20240719_1 and running the `get_idev_csr` test suite.

### CODE REVIEW

File: runtime/README.md [hunk 1]
```diff
@@ -794,6 +794,25 @@ Command Code: `0x4154_5348` ("ATSH")
...
+| chksum      | u32      | Checksum over other input arguments, computed by the caller. Little endian.  |
+
+*Table: `GET_IDEVID_CSR` output arguments*
+| **Name**      | **Type** | **Description**
+| --------      | -------- | ---------------
+| chksum        | u32      | Checksum over other output arguments, computed by Caliptra. Little endian. |
+| data\_size    | u32      | Length in bytes of the valid data in the data field.                       |
+| data          | u8[...]  | DER-encoded IDevID certificate signing request.                            |
+
+**Note**: The `GENERATE_IDEVID_CSR` Manufacturing State Flag **MUST** have been set to generate a CSR. Otherwise this command will return an empty data buffer.
```
Comment (clundin25) [inline runtime/README.md:814]: @jhand2 do you prefer an error code to an empty data buffer? The  flag gets cleared once the CSR is sent over the mailbox, so we either need to persist that the flag was set or inspect the CSR buffer, and map an empty buffer to an error code. 

Let me know what your preference is.

**Comment by clundin25** [comment]
@jhand2 I have a runtime implementation as well, but I need to refactor the runtime integration tests to allow setting the `GENERATE_CSR` flag.

**Review by mhatrevi — CHANGES_REQUESTED** [review]
Discussion pending on whether a ROM change is necessary.

**Comment by jhand2** [comment]
> Discussion pending on whether a ROM change is necessary.

https://github.com/chipsalliance/caliptra-sw/issues/1687 can't be done without a ROM change because no other code can wield IDevID to sign the CSR. But anyway, we can discuss the use case and if anyone is uncomfortable with the solution.

File: api/src/mailbox.rs [hunk 2]
```diff
@@ -976,6 +983,40 @@ impl Default for SetAuthManifestReq {
     }
 }
 
+// GET_IDEVID_CSR
+#[repr(C)]
+#[derive(Default, Debug, AsBytes, FromBytes, PartialEq, Eq)]
+pub struct GetIDevIDCSRReq {
```
Comment (jhand2) [inline api/src/mailbox.rs:989]: minor nit: The other APIs don't capitalize the acronyms. e.g. `GetIdevIdCsrReq`

File: drivers/src/idevid_csr.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+use zerocopy::{AsBytes, FromBytes};
+use zeroize::Zeroize;
+
+use crate::memory_layout;
+
+pub const MAX_CSR_SIZE: usize = 512;
+
+#[derive(FromBytes, AsBytes, Zeroize)]
+#[repr(C)]
+pub struct IDevIDCsr {
```
Comment (jhand2) [inline drivers/src/idevid_csr.rs:13]: nit: Should this struct go in the persistent data structure since it's really only intended to be used there?

File: drivers/src/idevid_csr.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    pub csr: [u8; MAX_CSR_SIZE],
+    pub csr_len: u32,
```
Comment (jhand2) [inline drivers/src/idevid_csr.rs:15]: I'm not positive, but it might be better for alignment to put `csr_len` as the first field.

File: drivers/src/idevid_csr.rs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+use zeroize::Zeroize;
+
+use crate::memory_layout;
+
+pub const MAX_CSR_SIZE: usize = 512;
+
+#[derive(FromBytes, AsBytes, Zeroize)]
+#[repr(C)]
+pub struct IDevIDCsr {
+    pub csr: [u8; MAX_CSR_SIZE],
```
Comment (jhand2) [inline drivers/src/idevid_csr.rs:14]: If we're going to have a get function, should the fields be non-public?

File: hw-model/src/lib.rs [hunk 6]
```diff
@@ -585,6 +585,15 @@ pub trait HwModel: SocManager {
...
             let mut cycles = 0;
             while !self.ready_for_fw() {
+                // If GENERATE_IDEVID_CSR was set then we need to clear cptra_dbg_manuf_service_reg
+                // once the CSR is ready to continue making progress.
+                //
+                // Generally the CSR should be read from the mailbox at this point, but to
+                // accommodate test cases that ignore the CSR mailbox, we will ignore it here.
+                if self.soc_ifc().cptra_flow_status().read().idevid_csr_ready() {
+                    self.soc_ifc().cptra_dbg_manuf_service_reg().write(|_| 0);
+                }
```
Comment (jhand2) [inline hw-model/src/lib.rs:595]: For tests which set the bit, I wonder if they should be responsible for clearing it (even if they're not going to read the CSR). I think it makes the expected behavior when reading the test a bit less magic.

File: rom/dev/src/fht.rs [hunk 7]
```diff
@@ -151,10 +151,7 @@ impl FhtDataStore {
 pub fn initialize_fht(env: &mut RomEnv) {
     let pdata = &env.persistent_data.get();
 
-    cprintln!(
-        "[fht] Storing FHT @ 0x{:08X}",
-        &pdata.fht as *const _ as usize
-    );
+    cprintln!("[fht] FHT @ 0x{:08X}", &pdata.fht as *const _ as usize);
```
Comment (jhand2) [inline rom/dev/src/fht.rs:154]: nit: Should we do a separate PR for cleaning up the log strings?

Another option would be to squash the actual feature into one commit and then split the log changes into a separate commit.

File: rom/dev/src/flow/cold_reset/fw_processor.rs [hunk 8]
```diff
@@ -306,6 +304,32 @@ impl FirmwareProcessor {
...
+                            /* CSR was never written to DCCM. This most likely means the
+                             * gen_idev_id_csr manufacturing flag was not set before booting into
+                             * ROM. */
+                            return Err(
+                                CaliptraError::FW_PROC_MAILBOX_GET_IDEV_CSR_UNPROVISIONED_CSR,
+                            );
+                        }
+
+                        resp.data_size = csr_persistent_mem.csr_len;
+                        resp.data[..resp.data_size as usize].copy_from_slice(csr);
```
Comment (jhand2) [inline rom/dev/src/flow/cold_reset/fw_processor.rs:328]: nit: Would be good to add a comment about why this is panic-free

File: rom/dev/src/flow/cold_reset/x509.rs [hunk 9]
```diff
@@ -16,29 +16,6 @@ use crate::cprintln;
 use crate::rom_env::RomEnv;
 use caliptra_drivers::*;
 
-/// Wrapper to hold certificate buffer and length
-pub struct Certificate<'a, const LEN: usize> {
```
Comment (jhand2) [inline rom/dev/src/flow/cold_reset/x509.rs:20]: Was this only used for the CSR

**Review by jhand2 — COMMENTED** [review]
Make it about half way throught. Will take another look later today or tomorrow.

As a general comment, I think moving the log stuff to a separate PR would help make this one a lot smaller.

File: runtime/tests/runtime_integration_tests/test_lms.rs [hunk 10]
```diff
@@ -838,7 +838,7 @@ fn test_lms_verify_cmd() {
 
 #[test]
 fn test_lms_verify_failure() {
-    let mut model = run_rt_test(None, None, None);
+    let mut model = run_rt_test(RuntimeTestArgs::default());
```
Comment (jhand2) [inline runtime/tests/runtime_integration_tests/test_lms.rs:841]: Also moving this refactor to a separate PR might be a good idea. Since the CSR change touches ROM, I think it would be good to keep this PR a bit smaller.

Comment (clundin25) [reply runtime/tests/runtime_integration_tests/test_lms.rs:841]: Opened https://github.com/chipsalliance/caliptra-sw/pull/1776

File: api/src/mailbox.rs (hunk 2 above)
Comment (clundin25) [reply api/src/mailbox.rs:989]: Done. I'll wait to push these changes until the UART logs PR is merged.

File: drivers/src/idevid_csr.rs (hunk 3 above)
Comment (clundin25) [reply drivers/src/idevid_csr.rs:13]: Done. I'll wait to push these changes until the UART logs PR is merged.

File: drivers/src/idevid_csr.rs (hunk 4 above)
Comment (clundin25) [reply drivers/src/idevid_csr.rs:15]: I think so too (at least similar structs do). I've made this change and will include it in the next rev.

File: hw-model/src/lib.rs (hunk 6 above)
Comment (clundin25) [reply hw-model/src/lib.rs:595]: We chatted offline and decided to keep this. 

It's unfortunate that this coupled logic is buried but the other options involve a log of duplicate logic or large refactor.

File: rom/dev/src/flow/cold_reset/x509.rs (hunk 9 above)
Comment (clundin25) [reply rom/dev/src/flow/cold_reset/x509.rs:20]: That's right.

File: drivers/src/idevid_csr.rs (hunk 5 above)
Comment (clundin25) [reply drivers/src/idevid_csr.rs:14]: Refactored this, but I will need to remove some more log statements to re-coop the bytes.

File: rom/dev/src/flow/cold_reset/fw_processor.rs (hunk 8 above)
Comment (clundin25) [reply rom/dev/src/flow/cold_reset/fw_processor.rs:328]: Added! Will be in next rev

File: drivers/src/persistent.rs [hunk 11]
```diff
@@ -44,6 +47,65 @@ pub type StashMeasurementArray = [MeasurementLogEntry; MEASUREMENT_MAX_COUNT];
 pub type AuthManifestImageMetadataList =
     [AuthManifestImageMetadata; AUTH_MANIFEST_IMAGE_METADATA_LIST_MAX_COUNT];
 
+#[derive(FromBytes, AsBytes, Zeroize)]
+#[repr(C)]
+pub struct IDevIDCsr {
```
Comment (jhand2) [inline drivers/src/persistent.rs:52]: nit: Change capitalization to match the other struct

File: error/src/lib.rs [hunk 12]
```diff
@@ -443,6 +443,10 @@ impl CaliptraError {
     pub const RUNTIME_CMD_RESERVED_PAUSER: CaliptraError = CaliptraError::new_const(0x000E004F);
     pub const RUNTIME_AUTH_AND_STASH_MEASUREMENT_DPE_ERROR: CaliptraError =
         CaliptraError::new_const(0x000E0050);
+    pub const RUNTIME_GET_IDEV_ID_UNPROVISIONED: CaliptraError =
+        CaliptraError::new_const(0x000E0051);
+    pub const RUNTIME_GET_IDEV_ID_UNSUPPORTED_ROM: CaliptraError =
```
Comment (jhand2) [inline error/src/lib.rs:448]: Maybe no action required, but can you double check that there's no overlap with the error codes in the `main-2.x` branch? Sam comment for all the errors in this file.

Ultimately I would like to pull the errors crate into a separate repo, but for now we may need to manually keep them in sync.

File: rom/dev/src/flow/cold_reset/fw_processor.rs [hunk 13]
```diff
@@ -304,6 +304,32 @@ impl FirmwareProcessor {
...
                     }
+                    CommandId::GET_IDV_CSR => {
+                        let mut request = MailboxReqHeader::default();
+                        Self::copy_req_verify_chksum(&mut txn, request.as_bytes_mut())?;
+
+                        let csr_persistent_mem = &persistent_data.idevid_csr;
+                        let mut resp = GetIdevIdCsrResp::default();
+
+                        if csr_persistent_mem.is_unprovisioned() {
+                            /* CSR was never written to DCCM. This most likely means the
```
Comment (jhand2) [inline rom/dev/src/flow/cold_reset/fw_processor.rs:315]: nit: are there other reasons it wouldn't be populated? if not we may want to make the language in the comment stronger.

File: rom/dev/src/flow/cold_reset/idev_id.rs [hunk 14]
```diff
@@ -264,37 +262,49 @@ impl InitDevIdLayer {
...
+        let mut result = Self::send_csr(env, &dev_id_csr);
+        if result.is_ok() {
+            result = Self::write_csr_to_peristent_storage(env, dev_id_csr);
+        }
+        csr_buf.zeroize();
 
         result
     }
 
+    fn write_csr_to_peristent_storage(env: &mut RomEnv, csr: IDevIDCsr) -> CaliptraResult<()> {
```
Comment (jhand2) [inline rom/dev/src/flow/cold_reset/idev_id.rs:294]: nit: are we confident this won't make a copy on the stack?

File: runtime/src/get_idv_csr.rs
Comment (jhand2) [inline runtime/src/get_idv_csr.rs:1]: nit: let's spell out `idev` in the file name

File: runtime/src/lib.rs [hunk 15]
```diff
@@ -226,6 +228,7 @@ fn handle_command(drivers: &mut Drivers) -> CaliptraResult<MboxStatusE> {
         CommandId::SHUTDOWN => FipsShutdownCmd::execute(drivers),
         CommandId::SET_AUTH_MANIFEST => SetAuthManifestCmd::execute(drivers, cmd_bytes),
         CommandId::AUTHORIZE_AND_STASH => AuthorizeAndStashCmd::execute(drivers, cmd_bytes),
+        CommandId::GET_IDV_CSR => GetIdevIdCsrCmd::execute(drivers, cmd_bytes),
```
Comment (jhand2) [inline runtime/src/lib.rs:231]: Same, I think it would be clearer to spell out idev

File: runtime/tests/runtime_integration_tests/test_get_idev_csr.rs [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+use caliptra_runtime::RtBootStatus;
+use openssl::x509::X509Req;
+use zerocopy::{AsBytes, FromBytes};
+
+use crate::common::{run_rt_test, RuntimeTestArgs};
+
+#[test]
+fn test_get_csr() {
+    let args = RuntimeTestArgs {
+        test_mfg_flags: Some(MfgFlags::GENERATE_IDEVID_CSR),
```
Comment (jhand2) [inline runtime/tests/runtime_integration_tests/test_get_idev_csr.rs:17]: nit: might be worth a comment that `run_rt_test` will clear this flag to allow RT to boot.

File: error/src/lib.rs (hunk 12 above)
Comment (clundin25) [reply error/src/lib.rs:448]: These error messages are not yet taken in `main-2.x`, so it should be okay assuming there is no open PR using them.

**Review by jhand2 — APPROVED** [review]

File: rom/dev/src/flow/cold_reset/idev_id.rs (hunk 14 above)
Comment (clundin25) [reply rom/dev/src/flow/cold_reset/idev_id.rs]: Changed to a Ref + Clone instead of the move.

We can follow up with a PR to assembly the CSR directly in the PersistentData memory slice if we want to save the stack space.

**Review by jhand2 — DISMISSED** [review]

**Review by mhatrevi — APPROVED** [review]

**Review by jhand2 — APPROVED** [review]
